File: src/stores/composer.ts
import { defineStore } from 'pinia'
import { computed, ref } from 'vue'
import { invoke } from '@tauri-apps/api/core'
import { listen } from '@tauri-apps/api/event'
import type { ComposerRelease, ComposerVersion } from '@/types/runtime'
import type { InstallProgress } from '@/types/binary'

// Keep in sync with `PROGRESS_EVENT` in src-tauri/src/services/binaries.rs
const PROGRESS_EVENT = 'binary://install-progress'

function errorMessage(e: unknown): string {
  if (typeof e === 'string') return e
  if (e instanceof Error) return e.message
  return 'Something went wrong.'
}

export const useComposerStore = defineStore('composer', () => {
  /** Kept for the pre-existing "is Composer ready at all" check (the
   *  requirements-check flow) — now backed by `services::composer`'s
   *  version-aware storage rather than a single fixed path. */
  const installed = ref(false)
  const installing = ref(false)
  const error = ref<string | null>(null)

  /** Versions currently on disk, and which one is active. */
  const versions = ref<ComposerVersion[]>([])
  const active = computed(() => versions.value.find((v) => v.active) ?? null)
  const switching = ref<string | null>(null)

  /** Stable releases Composer currently publishes. */
  const catalog = ref<ComposerRelease[]>([])
  const catalogLoading = ref(false)
  const catalogError = ref<string | null>(null)
  const installingVersion = ref<string | null>(null)

  const progress = ref<Record<string, InstallProgress>>({})
  listen<InstallProgress>(PROGRESS_EVENT, (event) => {
    progress.value[event.payload.id] = event.payload
  })

  async function fetchStatus() {
    installed.value = await invoke<boolean>('composer_installed')
  }

  /** Downloads the newest stable release if nothing is installed yet — the
   *  pre-existing entry point, kept for the requirements-check flow. */
  async function install() {
    installing.value = true
    error.value = null
    try {
      await invoke('install_composer')
      installed.value = true
    } catch (e) {
      error.value = errorMessage(e)
    } finally {
      installing.value = false
    }
  }

  async function fetchVersions() {
    versions.value = await invoke<ComposerVersion[]>('list_composer_versions')
  }

  async function setActive(id: string) {
    error.value = null
    switching.value = id
    try {
      versions.value = await invoke<ComposerVersion[]>('set_active_composer_version', { id })
    } catch (e) {
      error.value = errorMessage(e)
    } finally {
      switching.value = null
    }
  }

  async function fetchCatalog(refresh = false) {
    catalogLoading.value = true
    catalogError.value = null
    try {
      catalog.value = await invoke<ComposerRelease[]>('list_composer_catalog', { refresh })
    } catch (e) {
      catalogError.value = errorMessage(e)
      catalog.value = []
    } finally {
      catalogLoading.value = false
    }
  }

  async function installVersion(version: string) {
    installingVersion.value = version
    catalogError.value = null
    try {
      versions.value = await invoke<ComposerVersion[]>('install_composer_version', { version })
      installed.value = true
      const entry = catalog.value.find((r) => r.version === version)
      if (entry) entry.installed = true
      return true
    } catch (e) {
      catalogError.value = errorMessage(e)
      return false
    } finally {
      delete progress.value[`composer-${version}`]
      installingVersion.value = null
    }
  }

  function progressFor(version: string) {
    return progress.value[`composer-${version}`] ?? null
  }

  return {
    installed,
    installing,
    error,
    versions,
    active,
    switching,
    catalog,
    catalogLoading,
    catalogError,
    installingVersion,
    fetchStatus,
    install,
    fetchVersions,
    setActive,
    fetchCatalog,
    installVersion,
    progressFor,
  }
})

File: src-tauri/src/commands/composer.rs
//! Thin glue between the Switch page and `services::composer` /
//! `services::composer_catalog`.

use tauri::AppHandle;

use crate::services::composer::{self, ComposerVersionStatus};
use crate::services::composer_catalog::{self, ComposerRelease};
use crate::services::scaffold;
use crate::utils::error::AppError;

/// Whether at least one Composer version is installed — the label a project
/// template's requirements check and the Switch page both read.
#[tauri::command]
pub fn composer_installed() -> bool {
    scaffold::composer_installed()
}

/// Downloads the newest catalog version if nothing is installed yet.
/// Pre-existing entry point (predates per-version installs); scaffolding a
/// Laravel project does the same thing lazily on its own, so this exists
/// for a user who wants Composer ready ahead of time.
#[tauri::command]
pub async fn install_composer() -> Result<(), AppError> {
    scaffold::install_composer().await
}

/// The Composer versions currently on disk, and which one is active.
#[tauri::command]
pub fn list_composer_versions() -> Vec<ComposerVersionStatus> {
    composer::list()
}

#[tauri::command]
pub fn set_active_composer_version(id: String) -> Result<Vec<ComposerVersionStatus>, AppError> {
    composer::set_active(&id)
}

/// The versions Composer currently publishes as stable. Hits the network on
/// first use, then serves a cached copy until `refresh` is set.
#[tauri::command]
pub async fn list_composer_catalog(refresh: bool) -> Result<Vec<ComposerRelease>, AppError> {
    composer_catalog::list(refresh).await
}

#[tauri::command]
pub async fn install_composer_version(
    app: AppHandle,
    version: String,
) -> Result<Vec<ComposerVersionStatus>, AppError> {
    composer::install(&app, &version).await
}
